src/Test/Inspection/Core.hs

```diff
@@ -89,6 +88,26 @@ withLessDetail sdoc = withPprStyle defaultUserStyle sdoc
 type VarPair = (Var, Var)
 type VarPairSet = S.Set VarPair
 
+-- | Is there some ordering of the two lists which makes the equality function
+-- equal at every point?
+allEqual :: (a -> a -> Bool) -> [a] -> [a] -> Bool
+allEqual eq as bs
+  | length as /= length bs = False
+  | otherwise = evalState (go as) bs
+  where
+    go [] = gets null
+    go (a : as) = do
+      modify $ removeMatching (eq a)
+      go as
+
+-- | Remove the first element in the list matching the predicate.
+removeMatching :: (a -> Bool) -> [a] -> [a]
```

Comment:
```haskell
removeMatching :: (a -> Bool) -> [a] -> [a]
removeMatching _ [] = Nothing
removeMatching f (a : as)
  | f a = Just as
  | otherwise = (a :) <$> removeMatching f as
```

to short-circuit when there's nothing to remove?

Reply:
Oh, actually I remember that I wrote this once already: https://github.com/tibbe/unordered-containers/blob/efa43a2ab09dc6eb72893d12676a8e188cb4ca63/Data/HashMap/List.hs

Sorry for not remembering right from the beginning.

Reply:
Maybe I'm missing something obvious, but I don't see any circumstances in which this would help.

Reply:
```haskell
allEqual (==) "aaaaaaaa" "bbbbbbbb"
==> <state "bbbbbbbb" > go "aaaaaaaa" 
==> <state "bbbbbbbb" > go "aaaaaaa"
==> <state "bbbbbbbb" > go "aaaaaa"
==> <state "bbbbbbbb" > go "aaaaa"
==> <state "bbbbbbbb" > go "aaaa"
==> <state "bbbbbbbb" > go "aaa"
==> <state "bbbbbbbb" > go "aa"
==> <state "bbbbbbbb" > go "a"
==> <state "bbbbbbbb" > go ""
==> null "bbbbbbbb"
==> False
```

O(n^2) comparisons, where to fail in this case O(n) would be enough.

Reply:
Oh, nice catch, thanks!